### PR TITLE
[ONNX] NMS in ONNX

### PR DESCRIPTION
**Pull request by mbrookhart**
This PR adds ONNX support for NonMaxSuppression.

The ONNX API is a little odd, it roughly follows the TF op for combined_non_max_suppression. To implement it in tvm, I decided to go with relay while loops instead of writing a new TOPI kernel.

Getting it to work on CUDA was a pain, first, I needed to change a couple of input values from attributes to parameters, then the cuda kernels were out of spec and skipping tests. Implementing those tests and passing them required refactoring some of the kernels to bring them in spec, and then removing cuda threads in several places to get the CUDA kernel to return the correct results. 

I labeled this WIP because I'd like to spend more time trying to figure out how to speed up the cuda kernels, but I'd love ideas from anyone who's interested in looking at it.

@jroesch @jwfromm @zhiics @tkonolige @csullivan 

Because you've touched the cuda nms file, could you take a look @Laurawly @yongwww @kevinthesun ?

Thanks,
Matthew

### CODE REVIEW

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [inline python/tvm/topi/cuda/nms.py:646]: Could you show the performance benchmark on some popular OD model workloads with the modified nms.py?

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:646]: I've been attempted to get SSD-Mobilenet from the ONNX model zoo working, but I"m hitting other bugs. I'll post some perf metrics as soon as I can get that working.

File: python/tvm/relay/frontend/onnx.py
Comment (jwfromm) [inline python/tvm/relay/frontend/onnx.py:2256]: The comment and this variable's shape dont line up. Can you update the comment to describe all 4 dimensions of `onnx_out`?

File: python/tvm/relay/frontend/onnx.py
Comment (jwfromm) [inline python/tvm/relay/frontend/onnx.py:2273]: A comment explaining what this condition is checking would be helpful.

File: python/tvm/relay/frontend/onnx.py
Comment (jwfromm) [inline python/tvm/relay/frontend/onnx.py:2205]: A quick description of why we need three loops to implement onnx NMS would be helpful for readers trying to make sense of what's going on here.

File: tests/python/frontend/onnx/test_forward.py
Comment (jwfromm) [inline tests/python/frontend/onnx/test_forward.py:61]: Remove these commented lines before merging.

File: tests/python/frontend/onnx/test_forward.py
Comment (jwfromm) [inline tests/python/frontend/onnx/test_forward.py:3992]: What's the thought behind this change?

**Review by jwfromm — APPROVED** [review]

**Comment by jwfromm** [comment]
Overall LGTM. I think the documentation on the NMS onnx conversion could be more detailed since it's a little difficult to understand right now.

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:646]: I fixed a number of bugs in #6906 that allow me to run SSD-RN50 on CPU, but with those fixes, I'm hitting the isuse of cuda topk not supporting dynamic shapes, and the output of NMS is intrinsically dynamic. I'm back to trying to solve that problem before I can give you perf numbers.

File: tests/python/frontend/onnx/test_forward.py
Comment (mbrookhart) [reply tests/python/frontend/onnx/test_forward.py:3992]: :/ I don't think I did it, I wonder if it's a rule in autoformatting

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:646]: I'm still struggling to get full object detection models working on GPU due to the issues with dynamic topk in cuda, but while I work on that, I did some profiling of the unit test. With batch size = 2, num_anchors = 5, and num_classes = 2, I get the following performance on my 1070Ti:

Total Execution time: 900 microseconds
NMS time: 66 microseconds
get_valid_counts time: 15 microseconds.

it looks like the performance issues in the onnx kernel are all about the slicing/expand dims and loop bounds checking, here are the most expensive ops:

```
0                      fused_dyn_strided_slice_1        89.28450                                     
1                      fused_add_2                      87.05910                                     
2                      fused_less_min                   83.04660                                     
3                    fused_vision_non_max_suppression   66.43930                                     
4   fused_expand_dims_cast_concatenate_expand_dims...   53.65090                                     
5                      fused_reshape_concatenate        46.64780                                     
6                      fused_dyn_strided_slice          46.59800                                     
7                      fused_concatenate_2              33.06690                                     
8                      fused_dyn_broadcast_to_1         29.49310                                     
9                      fused_squeeze_concatenate        26.83250                                     
10                 fused_cast_expand_dims_concatenate   24.00320                                     
11  fused_concatenate_cast_like_less_cast_like_add...   23.53060                                     
12                     fused_dyn_strided_slice_2        23.39070                                     
13  fused_concatenate_cast_like_less_cast_like_add...   22.79010                                     
14                     fused_shape_of_1                 21.37450                                     
15                     fused_shape_of_2                 21.02070                                     
16                     fused_dyn_broadcast_to           18.43470                                     
17                     fused_vision_get_valid_counts    14.89870
```

cc @jwfromm @tkonolige

File: tests/python/frontend/onnx/test_forward.py
Comment (jroesch) [reply tests/python/frontend/onnx/test_forward.py:3986]: I think I did it, this means that Pytest will actually take command line flags when run as main, including `-k`, `-s`, etc.

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [inline python/tvm/topi/cuda/nms.py:133]: There's data race between line 129 and line 133 for `valid_count` for loading and writing.

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:646]: Could you leverage more in the changes for this file?

**Review by Laurawly — CHANGES_REQUESTED** [review]

File: python/tvm/topi/cuda/sort.py
Comment (Laurawly) [inline python/tvm/topi/cuda/sort.py:206]: Could you leverage more on the changes of this file based on performance results?

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:133]: There is definitely not a data race now, because I removed the threading :smile: But I think I see your point, this might be while I couldn't pass the test with threading on. I will investigate

Comment (tkonolige) [reply python/tvm/topi/cuda/nms.py:133]: Could you use atomic add here?

**Comment by mbrookhart** [comment]
#7005 re implemented some of the features in this PR, I'll rebase and try to reconcile.

**Comment by mbrookhart** [comment]
@Laurawly @kevinthesun I have rebased, but I was unable to get it passing tests with Yao's changes. I'm going back through the kernels one by one to see if I can get the faster versions to pass tests before attempting the ONNX integration.

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: atomic_add doesn't work with nvptx. That's a headache...

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: I do not have a race condition here because I'm not parallelizing over that axis. The issue with the current kernel is that it cannot implement the ONNX API because it cannot properly sort the output indices, and ONNX needs those outputs in the correct order to do some post-processing.

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [inline python/tvm/topi/cuda/nms.py:114]: This implementation of rearrange_indices_out_ir returns an undersized tensor in some case, I think the threading isn't quite right, but i haven't been able to fix.

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: I see. So we lose parallelism in `num_anchors` compared with the original implementation. Are we able to keep the level of parallelism while getting the correct output indices?

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: The test is disabled on master:
https://github.com/apache/tvm/blob/7a0d10ce8d4df27734949b944992e5c516f620d0/tests/python/relay/test_op_level5.py#L317-L321

The current kernel doesn't follow the spec, specifically, it uses the wrong index for the out_indices update:
```
out_indices[tid + k] = tid + k
```
instead of
```
out_indices[i * num_anchors + valid_count[i]] = j
```
Unfortunately, because valid_count depends on the iteration order of the loop on this axis, I don't see a way to get the correct index without serializing this axis.

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: If you have a way to get the right behavior with more threading, I'm very interested in hearing it.

File: python/tvm/topi/cuda/sort.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/sort.py:206]: I have not been able to pass tests with the kernels currently in master, so I can't compare apples to apples

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py]: I have not been able to pass tests with the kernels currently in master, so I can't compare apples to apples

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: To have maximum parallelism, we can do two iterations with the first marking all valid elements with flag 1, and the second doing a prefix sum to get the correct output index.

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: I think it takes three iterations and four extra output-allocated tensors

The first is what's currently in main, but outputting to a tmp_out and a tmp_out_indices, plus an is_valid tensor.

The second is a kernel to do cumsum on is_valid to get sort_indices

The third is to re-order the outputs based on the sort_indices.

It's a little messy. I'm reading gpugems on parallel cumsum, but I'll give it a try.

Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: Yes, that method was actually my first implementation of nms before PR #5339. And in that PR, what I improved was only getting the correct valid_count from nms_ir, and marking -1 to invalid outputs. And after nms_ir, I do a sort (in descending order), to get the sorted outputs as well as sorted output indicies. The second implementation is currently in main and much faster than the previous one.

Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: So for your case, if you want to pass the correct sorted output and output indices after `get_valid_counts_ir`, what you need to do is to do an argsort with the value `out` and indices `out_indices` which you get from `get_valid_counts_ir` from current main: 
```sort_tensor = argsort_thrust(
            score_tensor, valid_count=None, axis=1, is_ascend=False, dtype=valid_count_dtype)
```
And that is what did at the beginning of nms: https://github.com/apache/tvm/blob/main/python/tvm/topi/cuda/nms.py#L561

I don't pass in `valid_count` because I haven't reordered the invalid elements to the end of the output array yet but only mark those invalid elements as -1. But after sorting the whole array, the invalid elements will be sorted to the end of the array. So will their indices.

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: I don't see the sort?

It's faster, but it's not correct, that PR also disabled the relevant tests I need to ensure ONNX works properly.

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: I'll go back and look at what you had.

Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:197]: Do you happen to have a script to reproduce the benchmarks you had in that PR? I assume it was a TF import?

Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: It was actually a mxnet model and the model is a customized detection model, and `num_anchors` is really large. You can run this tutorial for a benchmark reference: https://tvm.apache.org/docs/tutorials/frontend/deploy_ssd_gluoncv.html

Comment (Laurawly) [reply python/tvm/topi/cuda/nms.py:197]: Actually, I mentioned in PR #5339 that the model is `ssd_resnet50_v1` with (1, 3, 512, 512) input shape and with Thrust turned on in build

**Comment by mbrookhart** [comment]
@Laurawly I took the mxnet example you provided and ran it with the debug runtime. It required a little bit of editing, APIs have changed slightly since that tutorial was written. Anyway, this is what I get on my 1070 TI with Thrust enabled.

main:
```
Ops                                                                                                     Time(us)    Time(%)  Shape
---                                                                                                     --------    -------  ----- 
fused_vision_non_max_suppression                                                                        139329.0    74.66    (1, 122640, 6)
fused_vision_get_valid_counts                                                                           124.255     0.067    (1, 122640, 6)     
```
this PR:
```   
fused_vision_get_valid_counts                                                                           46138.3    50.891   (1, 122640, 6)  
fused_vision_non_max_suppression                                                                        12319.8    13.589   (1, 122640, 6)
```

The get valid counts function slow down, but I'm actually seeing the total runtime of these ops decrease from 139.3ms to 58.5ms

My modifications to the example can be found here: https://gist.github.com/mbrookhart/df25427cbbfb3c73ed16be72c8525610

**Comment by Laurawly** [comment]
> @Laurawly I took the mxnet example you provided and ran it with the debug runtime. It required a little bit of editing, APIs have changed slightly since that tutorial was written. Anyway, this is what I get on my 1070 TI with Thrust enabled.
> 
> main:
> 
> ```
> Ops                                                                                                     Time(us)    Time(%)  Shape
> ---                                                                                                     --------    -------  ----- 
> fused_vision_non_max_suppression                                                                        139329.0    74.66    (1, 122640, 6)
> fused_vision_get_valid_counts                                                                           124.255     0.067    (1, 122640, 6)     
> ```
> 
> this PR:
> 
> ```
> fused_vision_get_valid_counts                                                                           46138.3    50.891   (1, 122640, 6)  
> fused_vision_non_max_suppression                                                                        12319.8    13.589   (1, 122640, 6)
> ```
> 
> The get valid counts function slow down, but I'm actually seeing the total runtime of these ops decrease from 139.3ms to 58.5ms
> 
> My modifications to the example can be found here: https://gist.github.com/mbrookhart/df25427cbbfb3c73ed16be72c8525610

The time measurement is not as fast as before because of this PR: https://github.com/apache/tvm/pull/7005. If you reverse back this one, you should get fairly good performance without any effect on the correctness.

**Comment by mbrookhart** [comment]
So, we seem to be at a bit of an impass. The implementation that is in master is not 100% correct, and it was even less correct before #7005. This PR is the best combination of correctness and performance we have a the moment. I would like to get it merged to have a correct implementation in place, but I would also like to find a way to speed up get valid counts, what I have here is not great.

I'm happy to spend the next few days trying to improve parallelism in get_valid_counts, but given that this is a performance improvement over main and it's more correct, I think we shouldn't block merging over the current performance, we could always do a second PR to improve performance once we have the unit tests in place

**Comment by mbrookhart** [comment]
@Laurawly I just rewrote get_valid_counts in the way you suggested. I still need to better parallelize the sum/conditional scan operation, but this takes it to:
```
Ops                                                                                                     Time(us)   Time(%)  Shape               Inputs  Outputs  
---                                                                                                     --------   -------  -----               ------  -------  
fused_vision_non_max_suppression                                                                        12105.9    25.723   (1, 122640, 6)      
fused_vision_get_valid_counts                                                                           3517.22    7.474    (1, 122640, 6)
```
I'm going to take another look at NMS before I try to parallelize the sum

**Comment by Laurawly** [comment]
> So, we seem to be at a bit of an impass. The implementation that is in master is not 100% correct, and it was even less correct before #7005. This PR is the best combination of correctness and performance we have a the moment. I would like to get it merged to have a correct implementation in place, but I would also like to find a way to speed up get valid counts, what I have here is not great.
> 
> I'm happy to spend the next few days trying to improve parallelism in get_valid_counts, but given that this is a performance improvement over main and it's more correct, I think we shouldn't block merging over the current performance, we could always do a second PR to improve performance once we have the unit tests in place

We have tested the correctness of ssd related detection models before PR #7005 and they are in the correctness margin for customers to ship. I agree on the part you updated where PR #7005 has changed for perf improvement. But I do suggest a perf comparison of your implementation and the one before PR #7005.

**Comment by mbrookhart** [comment]
> We have tested the correctness of ssd related detection models before PR #7005 and they are in the correctness margin for customers to ship.

This indicates that the kernel was correct for certain inputs, but not that it was correct overall.

**Comment by Laurawly** [comment]
> I'm happy to spend the next few days trying to improve parallelism in get_valid_counts, but given that this is a performance improvement over main and it's more correct, I think we shouldn't block merging over the current performance, we could always do a second PR to improve performance once we have the unit tests in place

Why don't you put the new implementation in a separate file say `nms_onnx.py` or a separate function. And we can see if we can merge it back once enough tests have passed to test the flakiness of the kernel and when it has better parallelism than using only `batch_size` number of threads.

**Comment by mbrookhart** [comment]
I really don't want the code fragmentation, especially when it's currently ~9x faster than main. I will keep pounding on speeding it up

**Comment by mbrookhart** [comment]
I found a combination of serial and threaded portions of NMS that combine to make it fast while still passing the tests:
```
fused_vision_get_valid_counts                                                                           3674.62    10.87    (1, 122640, 6)
fused_vision_non_max_suppression                                                                        402.791    1.191    (1, 122640, 6)
```
I'll go back to parallelizing the sum/scan on get_valid_counts tomorrow, but at this point, this is ~30x faster than main.

File: tests/python/relay/test_op_level5.py
Comment (masahi) [inline tests/python/relay/test_op_level5.py:396]: I think I removed this condition in https://github.com/apache/tvm/pull/7051 
Why is this diff here? Need rebase?

Comment (mbrookhart) [reply tests/python/relay/test_op_level5.py:396]: I haven't rebased since Monday, will fix

**Comment by mbrookhart** [comment]
@Laurawly Commit fd5ce6459, just before #7005, gets me this :/ I don't think the perf problem came from 7005
```
fused_vision_non_max_suppression                                                                        142188.0    76.13    (1, 122640, 6)
fused_vision_get_valid_counts                                                                           76.874      0.041    (1, 122640, 6)
```

File: python/tvm/topi/cuda/nms.py
Comment (Laurawly) [inline python/tvm/topi/cuda/nms.py:760]: Opencl doesn't have thrust lib, assert target is cuda here.

File: python/tvm/topi/cuda/sort.py
Comment (Laurawly) [inline python/tvm/topi/cuda/sort.py:108]: Doens't seem to make sort faster or more stable.

**Review by Laurawly — COMMENTED** [review]
Could you show perf comparison of this cuda nms implementation with this version of nms https://github.com/apache/tvm/blob/fdfc7eb8876278cbfd31f6d1d82ca75829a7aac4/python/tvm/topi/cuda/nms.py ?  Trying to understand where are the speedups coming from. I'll also run some benchmark from my end when I get some bandwidth.

File: python/tvm/topi/cuda/sort.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/sort.py:108]: It makes it support dynamic input shapes

Comment (mbrookhart) [reply python/tvm/topi/cuda/sort.py:108]: I'm working on a fuller rewrite of the sort kernel for stabilty in another branch, I understand why this isn't fast or stable, and I can do better for small inputs, but I need to rework things more to support full sized inputs.

File: python/tvm/topi/cuda/nms.py
Comment (mbrookhart) [reply python/tvm/topi/cuda/nms.py:760]: Good catch, thanks!

**Comment by mbrookhart** [comment]
@Laurawly fdfc7eb88 matches the other tests I've done on recent versions of the main branch, 131ms total time, 75% of runtime.

**Comment by Laurawly** [comment]
> @Laurawly [fdfc7eb](https://github.com/apache/tvm/commit/fdfc7eb8876278cbfd31f6d1d82ca75829a7aac4) matches the other tests I've done on recent versions of the main branch, 131ms total time, 75% of runtime.

This benchmark has a **large gap** between the one we tested before. We need to verify it on our end first. CC @zhiics @yidawang

**Comment by jwfromm** [comment]
@Laurawly I understand your concern about performance but I think @mbrookhart has provided pretty ample evidence that this code is **better than what we have in master in terms of both correctness and performance**. I think we should merge this and investigate improving performance even further in a follow-up PR. Since this PR itself causes no regressions, there's no need to make it even bigger with tons of additional optimization.

**Comment by Laurawly** [comment]
> @Laurawly I understand your concern about performance but I think @mbrookhart has provided pretty ample evidence that this code is **better than what we have in master in terms of both correctness and performance**. I think we should merge this and investigate improving performance even further in a follow-up PR. Since this PR itself causes no regressions, there's no need to make it even bigger with tons of additional optimization.

@zhiics is currently testing it. If it works for him, I’ll approve it.

**Review by Laurawly — APPROVED** [review]
Just verified with @zhiics , it works for us! Thanks for the contribution 😄

**Comment by mbrookhart** [comment]
Awesome, thank you for the help @Laurawly. The idea for speeding up get valid counts was especially helpful. I have an idea for parallelizing the last serial function there, but it follows a similar path to #7099, where I'm hitting some odd lowering issues. As soon as I can work those out, I'll submit a PR for parallel scan to further optimize get_valid_counts.

**Review by zhiics — APPROVED** [review]

**Comment by zhiics** [comment]
Thanks everyone!

**Pull request merged**
